src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs

```diff
@@ -0,0 +1,151 @@
+            {
+                _testAccountTokenHandle = null;
+                throw new Exception($"Failed to get SafeAccessTokenHandle for test account {TestAccountName}", new Win32Exception());
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_testAccountTokenHandle == null)
+                return;
+
+            _testAccountTokenHandle.Dispose();
+            _testAccountTokenHandle = null;
+
+            using (var principalCtx = new PrincipalContext(ContextType.Machine))
+            using (var userPrincipal = UserPrincipal.FindByIdentity(principalCtx, TestAccountName))
+            {
+                if (userPrincipal == null)
+                    throw new Exception($"Failed to get user principal to delete test account {TestAccountName}");
+
+                try
+                {
+                    userPrincipal.Delete();
+                }
+                catch (InvalidOperationException)
+                {
+                    // TODO: Investigate, it always throw this exception with "Can't delete object already deleted", but it actually deletes it.
+                }
+            }
+        }
+
+        // This method asserts if it impersonates the current identity, i.e.: it ensures that an actual impersonation happens
+        public void RunImpersonated(Action action)
+        {
+            using (WindowsIdentity serverIdentity = WindowsIdentity.GetCurrent())
+            {
+                WindowsIdentity.RunImpersonated(_testAccountTokenHandle, () =>
+                {
+                    using (WindowsIdentity clientIdentity = WindowsIdentity.GetCurrent())
+                        Assert.NotEqual(serverIdentity.Name, clientIdentity.Name);
+
+                    action();
+                });
+            }
+        }
+
+        [DllImport("advapi32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        private static extern bool LogonUser(string userName, string domain, string password, int logonType, int logonProvider, out SafeAccessTokenHandle safeAccessTokenHandle);
+    }
+
+    /// <summary>
+    /// Tests for the constructors for NamedPipeClientStream
+    /// </summary>
+    public class NamedPipeTest_CurrentUserOnly_Windows : NamedPipeTestBase, IClassFixture<TestAccountImpersonator>
+    {
+        private TestAccountImpersonator _testAccountImpersonator;
+
+        public NamedPipeTest_CurrentUserOnly_Windows(TestAccountImpersonator testAccountImpersonator)
+        {
+            _testAccountImpersonator = testAccountImpersonator;
+        }
+
+        [OuterLoop]
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsWindowsAndElevated))]
+        [InlineData(PipeOptions.None, PipeOptions.CurrentUserOnly)]
+        [InlineData(PipeOptions.CurrentUserOnly, PipeOptions.None)]
+        public void Connection_UnderDifferentUser_SingleSide_CurrentUserOnly_Fails(PipeOptions serverPipeOptions, PipeOptions clientPipeOptions)
```

Comment:
Is there a test where the server is the one running as the other user? I can't think why it would matter, but it would also be easy to do.

Reply:
I'm considering adding it when I add the identify validation on the Windows server side (right now it completely trusts the ACL).